### ISSUE TITLE
Send Later incompatible with PGP encryption (Error: encryption)

### ISSUE THREAD

**Issue by farfallosa** [Extended-Thunder/send-later#200]
Since the actual update, "Send later" shows an "error: encrypted" in drafts folder. I use OpenPG and in TB 78x OpgenPG is integrated and does automatically encrypt drafts - so e-mails set for send later, too.
Therefore "send later" does not work. :(

**Comment by jperryhouts** [comment]
When Enigmail was an independent plugin there was a way to turn off the "encrypt draft messages" option, but it seems that option is gone now that PGP is integrated into Thunderbird. There's really nothing I can do about that, Sorry.

You might consider opening a bug with Thunderbird if you want this fixed.

PS. This was always incompatible with Send Later, but the latest update is more proactive with telling you about it. Nothing changed in the latest update to make this incompatible.

**Comment by fabacab** [comment]
@farfallosa Did you follow through with @jperryhouts's suggestion to escalate this to the Mozilla Thunderbird dev team? I'm looking through their Bugzilla bug tracker now but can't see a relevant bug and want to ensure I'm not spamming them. The closest existing ticket I see that may be relevant is their [Bug ID 1168907](https://bugzilla.mozilla.org/show_bug.cgi?id=1168907), "Saved draft email is always encrypted when encryption is set to off for the composing email," but this is 6 years old.

**Comment by fabacab** [comment]
Meanwhile, @jperryhouts, would you be willing to share with us a little more about why encrypted drafts are incompatible with Send Later? At first blush, it seems to me like an encrypted draft should not be a showstopper for Send Later functionality, so long as the email being sent "later" does not require private key material (e.g., it is not being signed).

What is the technical block preventing Send Later from working with encrypted accounts? I ask because I have some familiarity with both JavaScript and RFC 4880, as well as a modicum of extension development experience, I'd like to see this functionality implemented, and I can potentially devote a modest number of hours to helping implement a solution. I'd like to help, if I can.

**Comment by jperryhouts** [comment]
This wouldn't be a problem if Thunderbird encrypted Draft messages using the recipient's public key, but the point of encrypting draft messages is that you can open them up again and continue editing them. So instead Thunderbird encrypts those messages with your own public key. That makes them both unreadable to your recipient, and also blocks Send Later from opening them without user interaction (note there *might* be some way for send later to open them if you don't have a passphrase on your private key, but that's not really a solution).

Two things will need to happen before this really has a workable solution. One is that Thunderbird needs to allow some programmatic access to the raw content of those messages, presumably just by allowing you to store them unencrypted in Drafts. The other is that Send Later needs to then be able to encrypt their contents before placing the message into the Outbox, when its scheduled time arrives. This will likely also require some accommodation on Thunderbird's side, since I don't know how easy it is for extensions to interact with the new builtin PGP system.

So far I haven't put much thought into the latter problem, because it's blocked until the first one is solved. But if you're interested in working with the Thunderbird devs on this I'd be happy to stay in the loop to ensure that whatever solution does get implemented in Thunderbird is aligned with Send Later's use case.

**Comment by farfallosa** [comment]
@fabacab thanks for nudging me. I just posted a new bug report at bugzilla (https://bugzilla.mozilla.org/show_bug.cgi?id=1681168)
It would be great if you could help fixing this problem. Because I don't think that TB will change their approach within a reasonable period of time, although this would be the best solution. (Not in Bugzilla but in forums this bug was reported several times.)
Note : The draft could be permanent decrypted by clicking into the subject line. But this does not have any effects on the date-time-stamp in the drafts column of a send later saved mail. @jperryhouts

**Comment by jperryhouts** [comment]
> Note : The draft could be permanent decrypted by clicking into the subject line.

Not sure I understand what you mean by this. When you view a draft message it's of course decrypted for you to see its contents, but the message as stored on disk / server is still encrypted. Unless I'm misunderstanding your comment.

**Comment by fabacab** [comment]
@jperryhouts:

> Thunderbird encrypts those [draft] messages with your own public key. That makes them both unreadable to your recipient, and also blocks Send Later from opening them without user interaction (note there might be some way for send later to open them if you don't have a passphrase on your private key, but that's not really a solution).

Hmm, I often send signed email via Thunderbird's built-in OpenPGP support and am not asked for my private key's passphrase. As I understand it, either the key or its passphrase is stored in memory (along with access to my other account passwords) after I provide my Thunderbird "primary password" after the program first starts up. I'll probably have to dig around in the codebase to see what's exactly happening, but this experience is why I'm surprised you'd need to prompt the user for their OpenPGP private key passphrase at all.

In any event, thanks for your work thus far and if I end up being able to contribute in some way I'll do so here.

@farfallosa:

> I don't think that TB will change their approach within a reasonable period of time

The Thunderbird team is actively working on OpenPGP support. Every recent release since built-in OpenPGP support has included changes and bugfixes to the way OpenPGP preferences are handled, so now's a good time to poke them about this sort of thing.

Also, depending on the complexity of the issue, we might not need them to do more than they've already done. For example, I just now noticed that there is a `autoEncryptDrafts` boolean exposed via [Thunderbird's Config Editor](https://support.mozilla.org/en-US/kb/config-editor). Toggling it doesn't seem to have an immediate effect, but maybe in combination with some other existing config option, we can find a working solution or at least a workaround to help unblock Jonathan.

**Comment by jperryhouts** [comment]
I assume that your private key is stored in memory, because the whole thing feels pretty transparent when you're using it, but I don't know that can be counted on to always be true. For example, plain gpg via the command line usually stores your private key in memory for some time period, then asks you to enter the passphrase again once that time period has elapsed.

There's also the problem that I'm not sure how to interact with the new PGP system via the extension API. The extension API is *very* lacking right now, and to my knowledge nothing exists to expose the encryption system to add-ons, even indirectly.

@fabacab Whatever investigating & work you're interested in taking on regarding this is very welcome. Let me know what you find.

**Comment by jperryhouts** [comment]
I'm going to re-open this issue because I'm positive this is not going to be the last I hear from users about it, and it will be useful to have somewhere to point them.

**Comment by farfallosa** [comment]
@jperryhouts As far as I can say, once decrypted, a draft remains decrypted. I.e. even after a restart of TB the draft remains decrypted and does not need to be decrypted again. Therefore I assume that the draft is saved plain after a first decryption.

Note: To mention, this happens even if the draft is neither signed nor encrypted.

**Comment by jperryhouts** [comment]
Use Ctrl+Shift+U to look at the message source, and then tell me that message is not encrypted...

**Comment by farfallosa** [comment]
It is …  I wonder how TB manages to keep a decrypted view over restarts even though the message itself remains encrypted.
